Source/WebKit/Shared/Cocoa/CoreIPCNSURLProtectionSpace.mm

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "CoreIPCNSURLProtectionSpace.h"
+
+#import "ArgumentCoders.h"
+
+#if PLATFORM(COCOA) && HAVE(WK_SECURE_CODING_NSURLPROTECTIONSPACE)
+
+@interface NSURLProtectionSpace (WKSecureCoding)
+- (NSDictionary *)_webKitPropertyListData;
+- (instancetype)_initWithWebKitPropertyListData:(NSDictionary *)plist;
+@end
+
+namespace WebKit {
+
+#define SET_OBJECT(NAME, CLASS, WRAPPER)    \
+    id NAME = dict[@#NAME];                 \
+    if ([NAME isKindOfClass:CLASS.class]) { \
+        auto var = WRAPPER(NAME);           \
+        m_data.NAME = WTFMove(var);         \
+    }
+
+CoreIPCNSURLProtectionSpace::CoreIPCNSURLProtectionSpace(NSURLProtectionSpace *ps)
+{
+    auto dict = [ps _webKitPropertyListData];
+
+    SET_OBJECT(host, NSString, CoreIPCString);
+
+    id port = dict[@"port"];
+    if ([port isKindOfClass:NSNumber.class])
+        m_data.port = [port unsignedShortValue];
+
+    id type = dict[@"type"];
+    if ([type isKindOfClass:NSNumber.class]) {
+        auto val = [type unsignedCharValue];
+        if (isValidEnum<WebCore::ProtectionSpaceBaseServerType>(val))
+            m_data.type = static_cast<WebCore::ProtectionSpaceBaseServerType>(val);
+    }
+
+    SET_OBJECT(realm, NSString, CoreIPCString);
+
+    id scheme = dict[@"scheme"];
+    if ([scheme isKindOfClass:NSNumber.class]) {
+        auto val = [scheme unsignedCharValue];
+        if (isValidEnum<WebCore::ProtectionSpaceBaseAuthenticationScheme>(val))
```

Comment:
I wonder how we are sure every scheme number must be converted to a valid enum. 
For example, if the framework makes a change to define a new scheme type with value 200 and WebKit does not update `ProtectionSpaceBaseAuthenticationScheme` accordingly, then the value will be dropped?

Reply:
Yes.  This is desired.  If they add a new scheme, they will coordinate with us.